menu.py

```diff
@@ -100,39 +121,65 @@ def key_press(self, key: Key) -> Menu | None:
                     selected_menu.call()
                     return_menu = self
                     if old_name != selected_menu.display_str:
-                        self.display.update_row(self.selected_on_display, selected_menu.display_str)
+                        self.display.update_row(
+                            self._display_row(self.selected),
+                            selected_menu.display_str,
+                            highlight=True
+                        )
             else:
                 return_menu = self
 
         return return_menu
 
     def redraw(self) -> None:
-        self.display.clear()
-        self.display.highlight_text(self.selected_on_display)
-        start_from = max(0, self.selected - self.selected_on_display)
-        for index in range(start_from, min(len(self.menu_elements), self.display.rows + start_from)):
-            self.display.push_back(self.menu_elements[index].display_str)
-
-class MenuRoot:
-    def __init__(self, root: Menu, display: BaseDisplayHandler = BaseDisplayHandler()) -> None:
-        self.root = root
-        self.current_menu: Menu | None = None
-        self.root.set_display(display)
-
-    def key_press(self, key: Key) -> bool:
-        """ @return true if menu closed """
-        if self.current_menu is None:
-            self.current_menu = self.root
-            self.current_menu.redraw()
-        else:
-            self.current_menu = self.current_menu.key_press(key)
-        return self.current_menu is None
+        display_str = [
+            menu.display_str
+            for menu in self.menu_elements[self.start_row : self.start_row + self.display.rows]
+        ]
+        logging.debug("start=%d\ndisplay_str=%s", self.start_row, str(display_str))
+        self.display.print_lines(display_str, highlight=self._display_row(self.selected))
 
 class Interface:
-    def __init__(self,*, menu: Menu, display: BaseDisplayHandler = BaseDisplayHandler()) -> None:
-        self.menu = MenuRoot(root=menu, display=display)
-        self.display = display
+    def __init__(self,*, menu: Menu, sensor_readings: SensorReadings, display: ScreenDisplay) -> None:
+        self._root_menu = menu
+        self._current_menu : Menu | None = None
+        self._display = display
+        self._lock = Lock()
+        self._root_menu.set_display(display)
+        self._readings = sensor_readings
+        self.show_data()
 
     def key_press(self, key: Key) -> None:
-        if self.menu.key_press(key):
-            self.display.clear()
+        """@brief react on pressed button"""
+        with self._lock:
+            if self._current_menu is None:
+                if key is not Key.CANCEL:
+                    self._current_menu = self._root_menu
+                    self._current_menu.redraw()
+            else:
+                self._current_menu = self._current_menu.key_press(key)
+                if self._current_menu is None:
```

Comment:
It can't be empty(I didn't overload __bool__) but I find this way to be more explicit and readable. This way you don't need to know if _current_menu can or can not be empty just if it's None.

Reply:
I see